src/components/loginRedirect/LoginRedirect.jsx

```diff
@@ -0,0 +1,238 @@
+  const { country, isWhitelisted } = details
+
+  const fetchLocationAndWhiteListedStatus = useCallback(async () => {
+    try {
+      const urlDetails = JSON.parse(Buffer.from(decodeURIComponent(params.login), 'base64').toString('ascii'))
+      setURLDetails({
+        vendorAddress: urlDetails.id,
+        vendorName: urlDetails.v,
+        requestedDetails: urlDetails.r,
+        vendorURL: urlDetails.web,
+        urlType: urlDetails.cbu ? 'cbu' : 'rdu',
+        url: urlDetails?.cbu ?? urlDetails?.rdu,
+      })
+      if (!urlDetails.web.includes(urlDetails?.cbu ?? urlDetails?.rdu)) {
+        setWarnings(data => ({ ...data, isWebDomainDifferent: true }))
+      }
+      if (!(await goodWallet.isVerified(urlDetails.id))) {
+        setWarnings(data => ({ ...data, isVendorWalletWhitelisted: false }))
+      }
+      const location = await API.getLocation()
+      const isWhitelisted = await goodWallet.isCitizen()
+      setDetails({ country: location?.name, isWhitelisted })
+    } catch (e) {
+      //nothing here
+    }
+  }, [])
+
+  useEffect(() => {
+    fetchLocationAndWhiteListedStatus()
+  }, [fetchLocationAndWhiteListedStatus])
+
+  const submit = ({ isRejected = false }) => {
+    let responseurlDetails = {}
+    if (!isRejected) {
+      const { requestedDetails } = urlDetails
+      responseurlDetails = {
+        a: { value: walletAddress, attestation: '' },
+        v: { value: isWhitelisted, attestation: '' },
+        ...(requestedDetails.includes('location') ? { I: { value: country, attestation: '' } } : {}),
+        ...(requestedDetails.includes('name') ? { n: { value: fullName, attestation: '' } } : {}),
+        ...(requestedDetails.includes('email') ? { e: { value: email, attestation: '' } } : {}),
+        ...(requestedDetails.includes('mobile') ? { m: { value: mobile, attestation: '' } } : {}),
+        nonce: { value: Date.now(), attestation: '' },
+      }
+      responseurlDetails.sig = goodWallet?.wallet?.eth?.accounts?.sign(
+        JSON.stringify(responseurlDetails),
+        goodWallet.accounts[0].privateKey,
+      ).signature
+    } else {
+      responseurlDetails = {
+        error: 'Authorization Denied',
+      }
+    }
+    if (isMobileNative) {
+      //app native functions
+      API.sendLoginVendorDetails(urlDetails.url, responseurlDetails)
+        .then(() => {
+          //minimize the app
+          BackHandler.exitApp()
+        })
+        .catch(() => {
+          //minimize the app
+          BackHandler.exitApp()
+        })
+    } else {
+      if (urlDetails.urlType === 'rdu') {
+        const encodedData = encodeURIComponent(btoa(JSON.stringify(responseurlDetails)))
+        window.location.href = `${urlDetails.url}?login=${encodedData}`
+      } else {
+        API.sendLoginVendorDetails(urlDetails.url, responseurlDetails)
+          .then(() => {
+            window.close()
+          })
+          .catch(() => {
+            window.close()
+          })
+      }
+    }
+  }
```

Comment:
```suggestion
   const sendResponse = useCallback(response => {
    const { url, urlType } = urlDetails

    if (urlType === 'rdu') {      
      openLink(`${url}?login=${encodeBase64Params(response)}`, '_self')
      return
    }

    API.sendLoginVendorDetails(url, response)
      .catch(e => log.warn('Error sending login vendor details', e.message, e))
      .finally(exitApp)
  }, [urlDetails])

  const deny = useCallback(() => {
    sendResponse({ error: 'Authorization Denied', })
  }, [sendResponse])
  
  const allow = useCallback(() => {
    const { wallet, accounts } = goodWallet
    const { accounts: signer } = wallet.eth
    const { privateKey } = first(accounts)
    const { requestedDetails } = urlDetails
    const detail = value => ({ value, attestation: '' })
    
    const map = {
      location: ['I', country],
      name: ['n', fullName],
      name: ['e', email],
      mobile: ['m', mobile],
    }
    
    const response = {
      a: detail(walletAddress),
      v: detail(isWhitelisted),
      noce: detail(Date.now())
    }

    forIn(response, ([property, value], requested) => {
      if (!requestedDetails.includes(requested)) {
        return
      }

      response[property] = detail(value)
    })
    
    
    const { signature } = signer.sign(JSON.stringify(response), privateKey)

    sendResponse({ ...response, sig: signature })
  }, [urlDetails, sendResponse, country, isWhitelisted, fullName, email, mobile, walletAddress])
```

Reply:
1. let's split it onto 3fn
2. i've adjusted the responseurlDetails composition, simplified it using props map and helper to produce value aith attestation
3. add import of openLink utility fn from lib/utils/linking. It also works on native, so redirect url should work now on Android too. If this isn't needed (web only) you could add `&& ! isMobileNative` to the `urlType === 'rdu'` condition

Reply:
resolved